test/tests/functional/pbs_que_resc_usage.py

```diff
@@ -0,0 +1,180 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+import time
+
+
+class TestQueRescUsage(TestFunctional):
+    """
+    Test built-in and custom resource behavior after
+    set/unset/assign/release resource on the queue
+    """
+    err_msg = 'Failed to get the expected resource value/result'
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        self.server.manager(MGR_CMD_SET, NODE, {
+                            'resources_available.ncpus': 5},
+                            self.mom.shortname)
+        a = {'queue_type': 'execution', 'enabled': True, 'started': True}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, id='new_q')
+
+    def test_built_in_resc(self):
+        """
+        Test built-in resource and check the assigned resource value
+        after unset and shouldn't set to any negative value
+        """
+        # Unset the resource after job completion
+        a = {'resources_available.ncpus': 4}
+        self.server.manager(MGR_CMD_SET, QUEUE, a, id='new_q')
+        j_attr = {ATTR_queue: 'new_q', 'Resource_List.select': '1:ncpus=3'}
+        j = Job(TEST_USER, j_attr)
+        j.set_sleep_time(20)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, jid)
+        q_status = self.server.status(
+            QUEUE, 'resources_assigned.ncpus', id='new_q', extend='f')
+        self.assertEqual(
+            int(q_status[0]['resources_assigned.ncpus']), 3, self.err_msg)
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid)
+        q_status = self.server.status(
+            QUEUE, 'resources_assigned.ncpus', id='new_q', extend='f')
+        self.assertEqual(
+            int(q_status[0]['resources_assigned.ncpus']), 0, self.err_msg)
+        self.server.manager(MGR_CMD_UNSET, QUEUE,
+                            'resources_available.ncpus', id='new_q')
+        q_status = self.server.status(QUEUE, id='new_q', extend='f')
+        self.assertNotIn('resources_available.ncpus',
+                         q_status[0].keys(), self.err_msg)
+        self.assertNotIn('resources_assigned.ncpus',
+                         q_status[0].keys(), self.err_msg)
+
```

Comment:
I thought the code specifically only unset resources_assigned on non-builtins.  ncpus is a builtin.

Reply:
It's for both built-in and non-built-in resources, the behavior is different only during unset.
Here we have added a built-in resource explicitly to the queue and unset the resource after job completion but if we unset the resource before job completion then code unset/removes only resources_assigned on non-built-ins.